libraries/WiFi/src/WiFiClient.cpp

```diff
@@ -294,6 +309,19 @@ int WiFiClient::connect(const char *host, uint16_t port)
 
 int WiFiClient::connect(const char *host, uint16_t port, int32_t timeout)
 {
+    if (WiFiGenericClass::getStatusBits() & WIFI_WANT_IP6_BIT) {
```

Comment:
There should be no need for this check here.

You only need one version of hostByName() (and inside that function it should return IPv6 if available, and IPv4 if not, i.e. the check should be inside hostByName()).

Reply:
To this I agree, it should be part of the function responsible to lookup an IP. But with the suggestion I made earlier, to allow to specifically set IPv4/IPv6 preference as an optional 3rd parameter.

Reply:
This links to the same argument. Legacy code supporting only IPv4 would break without this. We can't expect every Arduino code out there to magically support IPv6 if they hard-coded v4 address support everywhere (like we did in early Tasmota).

Reply:
> (like we did in early Tasmota).

... and current ESPEasy, just to name some random project :)

Reply:
There is already a solution for apps that only support IPv4, just simply don't set the flag CONFIG_LWIP_IPV6 and it will be backwards compatible.

Having the WIFI_WANT_IP6_BIT status bit seems redundant, or at least useless if IPv6 is turned off, so only relevant when you want IPv6 turned ON but deliberately the choose not to use it, which breaks the DNS abstraction.

For backwards compatibility, turn CONFIG_LWIP_IPV6 = OFF.

If you have IPv6 turned on, then you should be able to just say connect ("foo.example.com", ...) and the DNS system will abstract away what the real IP address is of foo, and including whether it is v4 or v6, or which is resolvable/connectable based on the current network?

The only rare situation seems to for diagnostics testing, where you want to test an IPv4 path on a dual stack network.

And even then hard coded addresses will still work, e.g. connect( "1.2.3.4", ..., will resolve to IPv4 1.2.3.4.

Reply:
Sorry but I don't agree. Changing CONFIG_LWIP_IPV6 = OFF is overly complex and requires to recompile the entire Arduino framework. This is far beyond the casual Arduino developer. Legacy dictates that code ignoring IPv6 should continue to work, and code willing to run with IPv6 has to enable it.

Reply:
That seems reasonable. If changing CONFIG_LWIP_IPV6 is difficult, then I understand the WIFI_WANT_IP6_BIT and think it would be a good solution.

But I would just want that one flag, and if that is all that is needed, that would be great.

i.e. have a single connect() method that resolves DNS and connects to either IPv4 if the WIFI_WANT_IP6_BIT flag is off, or to the best available (based on server, network, and device support) if it is on (i.e. IPv6 if available, otherwise also IPv4).

The same with a single hostByName() method that either has IPv4 only (if flag is off) or the best available supported (6 or 4) if on.

What I don't want is additional / different code beyond that single flag, i.e. a separate hostByName6() method ... otherwise that just causes confusion (setting WIFI_WANT_IP6_BIT would not work, as I would need to go around all over my code and change hostByName()).

(The only caveat to that is maybe you want a way to test an IPv4 route when IPv6 is available and supported, so for those diagnostics type operations have a separate hostByName(type = v4) or something method makes sense.)